lib/Basics/Guarded.h

```diff
@@ -0,0 +1,276 @@
+  [[nodiscard]] auto tryUnderLock(F&& callback) -> std::optional<Q>;
+  template <class F, class R = std::invoke_result_t<F, value_type&>,
+            class Q = std::conditional_t<std::is_void_v<R>, std::monostate, R>>
+  [[nodiscard]] auto tryUnderLock(F&& callback) const -> std::optional<Q>;
+
+  // get a copy of the value, made under the lock.
+  template <typename U = T, std::enable_if_t<std::is_copy_constructible_v<U>, int> = 0>
+  auto copy() const -> T;
+
+  // assign a new value using operator=, under the lock.
+  template <class U, std::enable_if_t<std::is_assignable_v<T, U>, int> = 0>
+  void assign(U&&);
+
+  using mutex_guard_type = MutexGuard<value_type, lock_type>;
+  using const_mutex_guard_type = MutexGuard<value_type const, lock_type>;
+
+  auto getLockedGuard() -> mutex_guard_type;
+  auto getLockedGuard() const -> const_mutex_guard_type;
+
+  auto tryLockedGuard() -> std::optional<MutexGuard<value_type, lock_type>>;
+  auto tryLockedGuard() const -> std::optional<MutexGuard<value_type const, lock_type>>;
+  // TODO add more types of "get guard" (like try or eventual)
+
+ private:
+  value_type _value;
+  mutable mutex_type _mutex;
+};
+
+template <class T, class M, template <class> class L>
+Guarded<T, M, L>::Guarded(T&& value) : _value{std::move(value)}, _mutex{} {}
+
+template <class T, class M, template <class> class L>
+template <typename... Args>
+Guarded<T, M, L>::Guarded(Args&&... args) : _value{std::forward<Args>(args)...}, _mutex{} {}
+
+template <class T, class M, template <class> class L>
+template <class F, class R>
+auto Guarded<T, M, L>::doUnderLock(F&& callback) -> R {
+  auto guard = lock_type(_mutex);
+  return std::invoke(std::forward<F>(callback), _value);
+}
+
+template <class T, class M, template <class> class L>
+template <class F, class R>
+auto Guarded<T, M, L>::doUnderLock(F&& callback) const -> R {
+  auto guard = lock_type(_mutex);
+  return std::invoke(std::forward<F>(callback), _value);
+}
+
+template <class T, class M, template <class> class L>
+template <class F, class R, class Q>
+auto Guarded<T, M, L>::tryUnderLock(F&& callback) -> std::optional<Q> {
+  auto guard = lock_type(_mutex, std::try_to_lock);
+
+  if (guard.owns_lock()) {
+    if constexpr (!std::is_void_v<R>) {
+      return std::forward<F>(callback)(_value);
+    } else {
+      std::forward<F>(callback)(_value);
+      return std::monostate{};
+    }
+  } else {
+    return std::nullopt;
+  }
+}
+
+template <class T, class M, template <class> class L>
+template <class F, class R, class Q>
+auto Guarded<T, M, L>::tryUnderLock(F&& callback) const -> std::optional<Q> {
+  auto guard = lock_type(_mutex, std::try_to_lock);
+
+  if (guard.owns_lock()) {
+    if constexpr (!std::is_void_v<R>) {
+      return std::forward<F>(callback)(_value);
+    } else {
+      std::forward<F>(callback)(_value);
+      return std::monostate{};
+    }
+  } else {
+    return std::nullopt;
+  }
+}
```

Comment:
Could we reduce the code duplication between the const and non-const implementations of tryUnderLock?

Reply:
Yes. :)